internal/discover/hooks.go

```diff
@@ -44,89 +67,128 @@ func (h *Hook) Hooks() ([]Hook, error) {
-	HookCreateSymlinks = HookName("create-symlinks")
-	// HookUpdateLDCache refers to the hook used to  Update the dynamic linker
-	// cache inside the directory path to be mounted into a container.
-	HookUpdateLDCache = HookName("update-ldcache")
-)
+type cdiHookCreator struct {
+	nvidiaCDIHookPath string
+	disabledHooks     map[HookName]bool
 
-// AllHooks maintains a future-proof list of all defined hooks.
-var AllHooks = []HookName{
-	HookEnableCudaCompat,
-	HookCreateSymlinks,
-	HookUpdateLDCache,
+	fixedArgs []string
 }
 
-type Option func(*CDIHook)
+// An allDisabledHookCreator is a HookCreator that does not create any hooks.
+type allDisabledHookCreator struct{}
 
-type CDIHook struct {
-	nvidiaCDIHookPath string
-	disabledHooks     disabledHooks
+// Create returns nil for all hooks for an allDisabledHookCreator.
+func (a *allDisabledHookCreator) Create(name HookName, args ...string) *Hook {
+	return nil
 }
 
+// A HookCreator defines an interface for creating discover hooks.
 type HookCreator interface {
 	Create(HookName, ...string) *Hook
 }
 
+// WithDisabledHooks sets the set of hooks that are disabled for the CDI hook creator.
+// This can be specified multiple times.
 func WithDisabledHooks(hooks ...HookName) Option {
-	return func(c *CDIHook) {
+	return func(c *cdiHookCreator) {
 		for _, hook := range hooks {
 			c.disabledHooks[hook] = true
 		}
 	}
 }
 
-func NewHookCreator(nvidiaCDIHookPath string, opts ...Option) HookCreator {
-	CDIHook := &CDIHook{
-		nvidiaCDIHookPath: nvidiaCDIHookPath,
-		disabledHooks:     disabledHooks{},
+// WithNVIDIACDIHookPath sets the path to the nvidia-cdi-hook binary.
+func WithNVIDIACDIHookPath(nvidiaCDIHookPath string) Option {
+	return func(c *cdiHookCreator) {
+		c.nvidiaCDIHookPath = nvidiaCDIHookPath
 	}
+}
 
+func NewHookCreator(opts ...Option) HookCreator {
+	cdiHookCreator := &cdiHookCreator{
+		nvidiaCDIHookPath: defaultNvidiaCDIHookPath,
+		disabledHooks:     make(map[HookName]bool),
+	}
 	for _, opt := range opts {
-		opt(CDIHook)
+		opt(cdiHookCreator)
+	}
+
+	if cdiHookCreator.disabledHooks[AllHooks] {
+		return &allDisabledHookCreator{}
 	}
 
-	return CDIHook
+	cdiHookCreator.fixedArgs = getFixedArgsForCDIHookCLI(cdiHookCreator.nvidiaCDIHookPath)
+
+	return cdiHookCreator
 }
 
-func (c CDIHook) Create(name HookName, args ...string) *Hook {
-	if c.disabledHooks[name] {
+// Create creates a new hook with the given name and arguments.
+// If a hook is disabled, a nil hook is returned.
+func (c cdiHookCreator) Create(name HookName, args ...string) *Hook {
+	if c.isDisabled(name, args...) {
 		return nil
 	}
 
-	if name == "create-symlinks" {
+	return &Hook{
+		Lifecycle: cdi.CreateContainerHook,
+		Path:      c.nvidiaCDIHookPath,
+		Args:      append(c.requiredArgs(name), c.transformArgs(name, args...)...),
+	}
+}
+
+// isDisabled checks if the specified hook name is disabled.
+func (c cdiHookCreator) isDisabled(name HookName, args ...string) bool {
```

Comment:
Moved the `isDisabled` check to a method and made this dependent on the `args` as well. This allows for the `create-symlinks` and `chmod` hooks to be handled more cleanly.